### PR TITLE
Update 3_train

### PR DESCRIPTION
**Pull request by AndyMcAliley**
This PR propagates two recent changes into `train.py`:

1. Validation data is split out during `2_process`.
2. Start dates and site IDs are saved with training and validation data.

## 1. Validation data is split out during `2_process`

This PR changes `train.py` to accommodate change 1 by loading `valid.npz` instead of splitting out validation set during training.

The function `get_data()` has been simplified by writing a new function: `get_sequence_dataset()`.
`get_data()` used to load a npz file, convert it into a `SequenceDataset` object, split that dataset into two (training and validation), and then send the two datasets to `get_data_loaders()`.
However, now there are two npz files to begin with, since `valid.npz` is created during `2_process`.
`get_sequence_dataset()` now handles the creation of a `SequenceDataset` object for one npz file.
All `get_data()` does now is call `get_sequence_dataset()` twice: once for `train.npz` and once for `valid.npz`, and send the resulting `SequenceDataset`s to `get_data_loaders()`.
The function `split()`, which used to split out validation data from training data, is no longer needed and has been deleted.

## 2. Start dates and site IDs are saved with training and validation data

This PR changes `train.py` to accommodate change 2 by saving the additional metadata with the training results.

The function `save_metadata()` has been modified to save start dates and site IDs for every sequence in both `train.npz` and `valid.npz`.

# How to run the code
These changes all concern the rule `train_model` in `3_train.smk`. So, to run this part of the pipeline, first copy `3_train/in/example_config.yaml` to `3_train/in/model_prep/1.yaml`. Then:
```
snakemake --cores all 3_train/out/model_prep/1/a_weights.pt
```
That should make `3_train/out/model_prep/1/1_process.yaml`, `3_train/out/model_prep/1/a_metadata.npz`, `3_train/out/model_prep/1/a_weights.pt`, and `3_train/out/model_prep/1/1_train.yaml`.

# How to review this PR

## Level of review requested

The main things I'd like reviewed are:

1. Do the pipeline additions and modifications make sense?
2. Have any errors been introduced?
3. Are the comments clear?

## Where in the code to focus

Anything that's been changed is fair game, but the majority of the changes are in `train.py`.

## Issues that will be addressed in upcoming PRs (so don't worry about them yet)
- Add early stopping
- Change directory structure to be more nested

### CODE REVIEW

File: 3_train/src/train.py [hunk 1]
```diff
@@ -381,14 +384,26 @@ def save_metadata(config, npz_filepath, save_filepath, overwrite=True):
...
+    metadata['train_site_ids'] = train_npz['site_ids']
+    metadata['valid_site_ids'] = valid_npz['site_ids']
+    # Remove start_dates and site_ids so that they don't get added to metadata twice
+    train_npz.files.remove('start_dates')
+    train_npz.files.remove('site_ids')
+    # The rest of the metadata in train_npz is identical to valid_npz
+    # So, we can add it from train_npz alone without any loss of info
+    for key in train_npz:
+        metadata[key] = train_npz[key]
     # Any duplicate keys get overwritten by the value in config
```
Comment (hcorson-dosch) [inline 3_train/src/train.py:407]:
```suggestion
    # Any duplicate keys get overwritten by the value in `config` (the train_config)
```

File: 3_train/src/train.py [hunk 2]
```diff
@@ -381,14 +384,26 @@ def save_metadata(config, npz_filepath, save_filepath, overwrite=True):
...
+    # The start dates and the site IDs are different for training and validation data
+    # Save start_dates and site_ids for both train and valid
+    metadata['train_start_dates'] = train_npz['start_dates']
+    metadata['valid_start_dates'] = valid_npz['start_dates']
+    metadata['train_site_ids'] = train_npz['site_ids']
+    metadata['valid_site_ids'] = valid_npz['site_ids']
+    # Remove start_dates and site_ids so that they don't get added to metadata twice
+    train_npz.files.remove('start_dates')
+    train_npz.files.remove('site_ids')
+    # The rest of the metadata in train_npz is identical to valid_npz
```
Comment (hcorson-dosch) [inline 3_train/src/train.py:403]:
```suggestion
    # The rest of the metadata in train_npz (training data means and standard 
    # deviations, process_config parameters) is identical to that in valid_npz
```

**Review by hcorson-dosch — APPROVED** [review]
This looks great, Andy. Thank you for the very clear explanation of your changes.
